### ISSUE TITLE
enforce @noloops constraint in all relevant mutation operations

### ISSUE THREAD

**Issue by hartig** [LiUGraphQL/woo.sh#47]
Relevant mutation operations are

*    create-object operations,
*    update-object operations,
*    create-edge operations, and
*    update-edge operations.

**Comment by epii-1** [comment]
This comment/question applies to #48, #49 and #50 too.

After looking a bit at graphql directives and apollo-servers, mostly from <https://www.apollographql.com/docs/apollo-server/schema/creating-directives/>, I see some problems with directives in combination transactions. If I understand it correctly directives are only applied either before the resolver is called (in the case of mutations and input fields) or when getting the result (in the case of queries and "normal" fields) and the directive of a specific field of the schema seems unreachable from, in our case, within `create` etc.. At least I have not been able to find a way to reach the directive of any field from within our `create` function, or any other function in `arangodb-tools` for that matter.  

This makes it, as of current, very complicated to make transactions compatible with some directives (for example @requiredForTarget), as these would require us to add aql-statements to the actual aql-transactions to be enforced correctly. If not, we run the risk of denying completely valid transactions, simply because one step is doing something that would temporarily brake the directive, even though later steps would fulfill the directive again.   

My question(s) therefor is
1. Have I misunderstood something about how directives works in apollo? If so can you please explain, or even better point me to a source that explains how.
2. Do we ignore the problem of transactions being declined to early, for the directives preventing them?
3. Do we force access to the directives for the functions in `arangodb-tools`? This can be done for example letting them parse api-schema, to get the directives that way, or by storing a map of (type-field, [directives]) pairs somewhere.
4. Is there some other way to implement directives, without having access to them during the creation of the aql-transactions, that I don't see?

@hartig

**Comment by hartig** [comment]
Directives are not "applied" at all. Instead, they are just additional annotations in the schema and they may be interpreted whenever the schema is processed or whenever some query or mutation operation uses a schema element (e.g., a field) that is annotated with a directive in the schema.

Now, the reason why you cannot get access to our constraints-defining directives within arangodb-tools.js is simple: they are not part of the schema anymore at this point. Remember that the generated GraphQL server (and arangodb-tools.js is part of this generated server) has access only to the API schema, not to the DB schema from which the API schema was generated. Now, while our constraints-defining directives may be used in the DB schema, our tool that generates the API schema (graphql-api-generator) currently does not carry over these annotations into the API schema. In a way, I also don't feel they belong there.

However, to be able to use them in arangodb-tools.js they may have to be copied over into the API schema. This needs to be done by graphql-api-generator. Nonetheless, I would not like them to be copied into the object types (which are a query-related aspect of the API schema), but instead they should be copied into the corresponding input types.

For instance, consider the following type as defined in a DB schema.
```graphql
type Person {
  name: String!  @required
  bestFriend: Person  @noloops
}
```
For this type, our graphql-api-generator puts the following object type into the API schema.
```graphql
type Person {
  name: String!
  bestFriend: Person
}
```
Additionally, the following two input types will be added:
```graphql
type _InputToCreatePerson {
  name: String!
  bestFriend: _InputToConnectBestFriendOfPerson
}
type _InputToUpdatePerson {
  name: String
  bestFriend: _InputToConnectBestFriendOfPerson
}
```
Now, my proposal is to copy the constraints-defining directives into these input types as follows:
```graphql
type _InputToCreatePerson {
  name: String!
  bestFriend: _InputToConnectBestFriendOfPerson  @noloops
}
type _InputToUpdatePerson {
  name: String
  bestFriend: _InputToConnectBestFriendOfPerson  @noloops
}
```
Then, you can use these annotations in arangodb-tools.js. For instance, in the `create` function you have the `info` argument which allows you to access the API schema via `info.schema`, and from there you should be able to find the necessary input file, its fields, and their directives.

For copying the directives over from the types in the DB schema to the corresponding **input** types in the API schema, you need to extend our graphql-api-generator tool. If you look at [line 257 of utils.py](https://github.com/LiUGraphQL/woo.sh/blob/b62c9e24b52f04425185c469e5cd249330d24421/graphql-api-generator/utils/utils.py#L257), you can get an idea of how to access the directives in Python.

/cc @keski

**Comment by epii-1** [comment]
Thanks for the answer.

> Directives are not "applied" at all. Instead, they are just additional annotations in the schema and they may be interpreted whenever the schema is processed or whenever some query or mutation operation uses a schema element (e.g., a field) that is annotated with a directive in the schema.

"Applied" might have been a bad choice of word. And I do agree with everything you said here.

However, I'm already aware of the rest of your comment. I even force inserted some directives into `api-schema` just to have something to test on. For example
 ```
type AudioFile implements MediaFile {
  path: String! @required
  ...
}
```

And yet I can not for the life of me seem to actually access the directive(s) of `info.schema.getType('AudioFile').getFields()['path']`. There is no `.directive`, `.directives`, `.getDirectives()`, `.required`, `.@required`, nor anything similar that I seem to find.

The only real usage of directives in apollo-servers I find are either the link I gave earlier, or examples, all of them being very similar to that link.

At the same time, this feels really weird; the directives should really be reachable from `info.schema`.

@hartig

**Comment by hartig** [comment]
Ah, I see. I did a quick search and found this: https://stackoverflow.com/questions/53974759/graphql-apollo-can-you-access-directives-within-resolve-function

So, perhaps, directives are simply not meant to be accessed from within resolvers :-(
But still, maybe it is possible to access them somehow?

Another alternative might be add the DB schema as a separate object inside the context object that is being passed around?

**Comment by epii-1** [comment]
We basically reached the same conclusion then.
To me it sounds like directives are more ment to be used as something resembling decorators.

Good to know that I'm not *that* stupid at least :)

As you are, ever so slightly, implying, there are ways around this. The question now just becomes deciding how "hacky" we want our solution to be.

Another possibility would be to pass the `typeDefs` to `tools.init` from `server`, and then parse this once into some format similar to my point 3 from earlier. We would then have no need passing this around as it would always be reachable from any function in `tools`.

**Comment by hartig** [comment]
At this point, I also don't know what an elegant solution could be.

> The question now just becomes deciding how "hacky" we want our solution to be.

The less "hacky" the better ;-)

When thinking about your proposal (pass the `typeDefs` to `tools.init` from `server`), I notice that `server.schema` is passed to `tools.init` (resp. `driver.init` as it is called now). Is this already a representation of the API schema from which the directives cannot be accessed anymore? If not, then we could simply assign this `schema` object to a global variable within inside ArangoDB driver.js and then use this variable in the relevant functions such as `create` and `update`.

On the other hand, if this `schema` object does not contain the directives anymore, I agree that `typeDefs` may be parsed again separately and the result may be passed into `driver.init`.

**Comment by keski** [comment]
It seems the directives have to be accessed via the field's `astNode` (see issue https://github.com/graphql-python/graphql-core/issues/65).  Given the type:
```graphql
type AudioFile implements MediaFile {
    path: String! @required
    ...
}
```
we would therefore access the directives associated with the `path` field using:
```javascript
info.schema.getType('AudioFile').getFields()['path'].astNode.directives;
```
which would return an object like:
```javascript
[
  {
    kind: 'Directive',
    name: { kind: 'Name', value: 'noloops', loc: [ ... ] },
    arguments: [],
    loc: { start: ..., end: ... }
  }
]
```

I couldn't find any convenience method for listing these though, so we probably have to roll our own util function.

**Comment by epii-1** [comment]
So the directives are accessible after all then. That simplifies a lot. Thanks for finding it Robin!

Quickly answering Olafs question: I think the `server.schema` is the same thing as `info.schema`. But I did not really go into details as now we have a proper solution for my problems anyway.

However, my initial questions here raises another question: Should we allow handling of directives in a similar fashion to <https://www.apollographql.com/docs/apollo-server/schema/creating-directives/> ? (Should probably be its own issue) We will probably not use it ourselves, but it would be nice to give users the option.

**Comment by hartig** [comment]
Great find Robin. Thanks!

Leif, to your extra question: At the moment, I don't see a need to worry about user-defined directives.

**Comment by hartig** [comment]
Addressed in PR #54 and to be followed up in #61 and #62
